File: Python3/2536_increment_submatrices_by_one.py
# https://leetcode.cn/problems/increment-submatrices-by-one/

class Solution:
    def rangeAddQueries(self, n: int, queries):
        """
        Inputs:
            n: int
            queries: list[list[int]]
        Outputs:
            res: list[list[int]]
        """
        diff = [[0 for _ in range(n+1)] for _ in range(n+1)]
        for q in queries:
            i1, j1, i2, j2 = q[0], q[1], q[2], q[3]
            diff[i1][j1] += 1
            diff[i2+1][j1] -= 1
            diff[i1][j2+1] -= 1
            diff[i2+1][j2+1] += 1
        for i in range(n):
            for j in range(1, n):
                diff[i][j] += diff[i][j-1]
        for i in range(1, n):
            for j in range(n):
                diff[i][j] += diff[i-1][j]

        res = [[0 for _ in range(n)] for _ in range(n)]
        for i in range(n):
            for j in range(n):
                res[i][j] = diff[i][j]
        return res

File: Cpp/2536_increment_submatrices_by_one.cpp
// https://leetcode.cn/problems/increment-submatrices-by-one/

#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
    vector<vector<int>> rangeAddQueries(int n, vector<vector<int>>& queries) {
        vector<vector<int>> diff(n+1, vector<int>(n+1, 0));
        for (const auto &q: queries) {
            int i1 = q[0], j1 = q[1], i2 = q[2], j2 = q[3];
            diff[i1][j1] += 1;
            diff[i2+1][j1] -= 1;
            diff[i1][j2+1] -= 1;
            diff[i2+1][j2+1] += 1;
        }
        for (int i = 1; i < n; ++i) {
            for (int j = 0; j < n; ++j) {
                diff[i][j] += diff[i-1][j];
            }
        }
        for (int i = 0; i < n; ++i) {
            for (int j = 1; j < n; ++j) {
                diff[i][j] += diff[i][j-1];
            }
        }

        vector<vector<int>> res(n, vector<int>(n, 0));
        for (int i = 0; i < n; ++i) {
            for (int j = 0; j < n; ++j) {
                res[i][j] = diff[i][j];
            }
        }
        return res;
    }
};

int main() {
    Solution sol;
    int n = 3;
    vector<vector<int>> queries = {{1,1,2,2}, {0,0,1,1}};
    auto res = sol.rangeAddQueries(n, queries);
    for (const auto &row: res) {
        for (const int &val: row) cout << val << ", ";
        cout << '\n';
    }
    return 0;
}